nanoCAT/recipes/coordination_number.py

```diff
@@ -0,0 +1,129 @@
+
+    The combinations, by definition, do not contain any atom pairs where ``at1.symbol == at2.symbol``
+    """
+    x, y = [], []
+    symbol_combinations = combinations(idx_dict.keys(), r=2)
+    for symbol1, symbol2 in symbol_combinations:
+        idx1 = idx_dict[symbol1]
+        idx2 = idx_dict[symbol2]
+        _x, _y = np.meshgrid(idx1, idx2)
+        x += _x.ravel().tolist()
+        y += _y.ravel().tolist()
+
+    x = np.fromiter(x, dtype=int, count=len(x))
+    y = np.fromiter(y, dtype=int, count=len(y))
+    return x, y
+
+
+def upper_dist(xyz: np.ndarray, x: np.ndarray, y: np.ndarray) -> np.ndarray:
+    """Construct the upper distance matrix."""
+    shape = len(xyz), len(xyz)
+    dist = np.full(shape, fill_value=np.nan)
+    dist[x, y] = np.linalg.norm(xyz[x] - xyz[y], axis=1)
+    return dist
+
+
+def radius_inner(mol: Molecule, idx_dict: Dict[str, np.ndarray]) -> Dict[str, float]:
+    """Search the threshold radius of the inner coordination shell for each atom type.
+
+    This radius is calculated as the distance with the first neighbors:
+    in case of heterogeneous coordinations, only the closest neighbors are considered.
+    """
+    d_inner = {}
+    symbol_combinations = combinations(idx_dict.keys(), r=2)
+    for symbol1, symbol2 in symbol_combinations:
+        d_pair = guess_core_core_dist(mol, (symbol1, symbol2))
+        if d_pair < d_inner.get(symbol1, np.inf):
+            d_inner[symbol1] = d_pair
+        if d_pair < d_inner.get(symbol2, np.inf):
+            d_inner[symbol2] = d_pair
+    return d_inner
+
+
+def coordination_inner(dist: np.ndarray, d_inner: Dict[str, float],
+                       idx_dict: Dict[str, np.ndarray], length: int) -> NestedDict:
+    """Calculate the coordination number relative to the inner shell (first neighbors)."""
+    coord_inner = np.empty(length, int)
+    for k, v in idx_dict.items():
+        a, b = np.where(dist <= d_inner[k])
+        count = np.bincount(a, minlength=length) + np.bincount(b, minlength=length)
+        coord_inner[v] = count[v]
+    return coord_inner
+
+
+def coordination_outer(dist: np.ndarray, d_outer: float, length: int) -> np.ndarray:
+    """Calculate the coordination number relative to the outer shell."""
+    a, b = np.where(dist <= d_outer)
+    coord_outer = np.bincount(a, minlength=length) + np.bincount(b, minlength=length)
+    return coord_outer
+
+
+def map_coordination(coord: np.ndarray, idx_dict: Dict[str, np.ndarray]) -> NestedDict:
+    """Map atoms according to their atomic symbols and coordination number.
+
+    Construct a nested dictionary {'atom_type1': {coord1: [indices], ...}, ...}
+    """
+    cn_dict = {}
+    for k, v in idx_dict.items():
+        cn = coord[v]
+        mapping = {i: (v[cn == i] + 1).tolist() for i in np.unique(cn)}
+        cn_dict[k] = mapping
+    return cn_dict
+
+
+def coordination_number(mol: Molecule, shell: str = 'inner', d_outer: float = None) -> NestedDict:
```

Comment:
Btw, the annotation for a value which can be ``None`` or an instance of ``float`` is ``Optional[float]``.